napari/components/command_palette/_api.py

```diff
@@ -0,0 +1,235 @@
+from __future__ import annotations
+
+import inspect
+import weakref
+from typing import TYPE_CHECKING, Callable, TypeVar, overload
+
+from napari.components.command_palette._components import Command
+
+if TYPE_CHECKING:
+    from qtpy import QtWidgets as QtW
+
+    from napari._qt.qt_main_window import _QtMainWindow
+    from napari._qt.widgets.qt_command_palette import QCommandPalette
+
+    _WVDict = weakref.WeakValueDictionary[int, _QtMainWindow]
+
+_F = TypeVar("_F", bound=Callable)
+
+
+def _always_true(*_) -> bool:
+    return True
+
+
+@inspect.signature
+def register_with_func(
+    func: Callable,
+    title: str | None = None,
+    desc: str | None = None,
+    tooltip: str | None = None,
+    when: Callable[..., bool] | None = None,
+):
+    """Template function to provide signature to register() with 'func' argument."""
+
+
+@inspect.signature
+def register_without_func(
+    title: str | None = None,
+    desc: str | None = None,
+    tooltip: str | None = None,
+    when: Callable[..., bool] | None = None,
+):
+    """Template function to provide signature to register() without 'func' argument."""
+
+
+class CommandPalette:
+    """The command palette interface."""
+
+    def __init__(self, name: str) -> None:
+        self._commands: list[Command] = []
+        self._parent_to_palette_map: dict[int, QCommandPalette] = {}
+        self._palette_to_parent_map: _WVDict = weakref.WeakValueDictionary()
+        self._name = name
+
+    @property
+    def commands(self) -> list[Command]:
+        """List of all the commands."""
+        return self._commands.copy()
+
+    @overload
+    def register(
+        self,
+        func: _F,
+        title: str | None,
+        desc: str | None = None,
+        tooltip: str | None = None,
+        when: Callable[..., bool] = None,
+    ) -> _F:
+        ...
+
+    @overload
+    def register(
+        self,
+        title: str | None,
+        desc: str | None = None,
+        tooltip: str | None = None,
+        when: Callable[..., bool] = None,
```

Comment:
I will need to double check this but for some reason I remember `app-model` commands having both `when` and `enablement` clauses and I feel like `when` is whether is shows up in the menu at all and `enablement` is whether it's greyed out or not. In which case we'd probably be wanting `enablement` here? Again I need to check on this though.